### PR TITLE
Include workflow inputs and logs in output directory

### PR DESCRIPTION
**Pull request by johnbradley**
Changes the lando_worker to create an output directory containing results, workflow input files, logs, and a report. This directory is currently uploaded to the user specified DukeDS project.

Directory layout:
```
upload_directory/               # this is a user specified name
    Bespin-Report.txt           # describes contents of the upload_directory
    results/
        ...output files from workflow
    logs/
        cwltool-output.json     # stdout from cwl-runner - json job results
        cwltool-output.log      # stderr from cwl-runner
        job-data.json           # non-cwl job data used to create Bespin-Report.txt
    workflow/
        workflow.cwl            # cwl workflow we will run
        workflow.yml            # job order input file
```

The report is really bare bones and not meant to be a final product.

### CODE REVIEW

File: lando/worker/cwlworkflow.py [hunk 1]
```diff
@@ -3,9 +3,90 @@
...
+REPORT_FILENAME = 'Bespin-Report.txt'
+
+LOGS_DIRECTORY = 'logs'
+JOB_STDOUT_FILENAME = 'cwltool-output.json'
+JOB_STDERR_FILENAME = 'cwltool-output.log'
+JOB_DATA_FILENAME = 'job-data.json'
+
+WORKFLOW_DIRECTORY = 'workflow'
+WORKFLOW_FILENAME = 'workflow.cwl'
+JOB_ORDER_FILENAME = 'workflow.yml'
```
Comment (dleehr) [inline lando/worker/cwlworkflow.py]: Files named `workflow.cwl` and `workflow.yml` aren't very descriptive. Even as someone with CWL experience I find this confusing - At a glance I can't tell which is the workflow.

Why not carry forward a descriptive name for the workflow (currently `rnaseq-pt1.cwl`) and something like `job-34-input.yml` for the job input file?

Also, is the job order YAML or JSON? I thought we stored it as JSON in bespin-api

File: lando/worker/cwlworkflow.py [hunk 2]
```diff
@@ -21,82 +102,176 @@ def __init__(self, job_id, working_directory, output_directory, cwl_base_command
...
+    def total_runtime_str(self):
+        """
+        Returns a string describing how long the job took.
+        :return: str: "<number> minutes"
+        """
+        elapsed_seconds = (self.finished - self.started).total_seconds()
+        return "{} minutes".format(elapsed_seconds / 60)
+
+
+class ResultsDirectory(object):
```
Comment (dleehr) [inline lando/worker/cwlworkflow.py]: Awesome!

Some **suggestions**, open for **discussion** :)

- Let's do away with the user-specified directory naming. Instead, generate a new project for every job, and name it based on:
  - Bespin
  - workflow name / version
  - job name
  - job creation date
- I think this fits better into other DukeDS structures (project transfers, downstream user permissions, etc), but I'm interested to hear alternate perspectives, or if we should wait and see.

Suggestions to the upload directory contents:

- Bespin-Report.txt: rename to README
- results: rename to output
- workflow: rename to scripts and add local README on how to run this

Informed by https://reproducible-science-curriculum.github.io/organization-RR-Jupyter/

File: lando/worker/cwlworkflow.py (hunk 1 above)
Comment (johnbradley) [inline lando/worker/cwlworkflow.py]: CWL supports both formats. The CWL guide shows everything as YAML (JSON is valid YAML). We are storing it as JSON in the database field when generating it from a questionnaire.

Comment (johnbradley) [inline lando/worker/cwlworkflow.py]: I'm fine with changing to the names you specify.

File: lando/worker/cwlworkflow.py (hunk 2 above)
Comment (johnbradley) [inline lando/worker/cwlworkflow.py]: Agreed on user specifying output project and directory name.
Transferring the project also keeps the user from seeing partial data sets or uploads that failed to complete.

File: lando/worker/cwlworkflow.py (hunk 1 above)
Comment (dleehr) [inline lando/worker/cwlworkflow.py]: > JSON is valid YAML

While that's a nice feature of the YAML parser, it's not obvious to people (users, developers) unless they know the trick or are experts with both formats. Can we stick with JSON?

Comment (johnbradley) [inline lando/worker/cwlworkflow.py]: Sure.

File: lando/worker/cwlworkflow.py (hunk 2 above)
Comment (johnbradley) [inline lando/worker/cwlworkflow.py]: One down side is you cannot move a folder/file between projects in DukeDS.

**Comment by johnbradley** [comment]
This PR requires https://github.com/Duke-GCB/bespin-api/pull/30 and https://github.com/Duke-GCB/lando-messaging/pull/4.

**Pull request merged**